### PR TITLE
Pensieve branch

### PR DESCRIPTION
**Pull request by francisyyan**
For code review only

### CODE REVIEW

File: src/abr/pensieve.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#include "pensieve.hh"
+#include "ws_client.hh"
+
+using namespace std;
+
+//TODO: Reference already existing put_field()
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: Idea for this TODO: we can add two files such as `serialization.hh` and `serialization.cc` in `src/net`, and make every file using `put_field`/`get_uint16` reference to `serialization.hh`.

File: src/abr/pensieve.hh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  size_t next_format_ {};
+  static FileDescriptor open_ipc() {
+    std::cout << "removing old ipc" << std::endl;
+    remove("/tmp/pensieve");
+    IPCSocket sock;
+    sock.bind("/tmp/pensieve");
+    // TODO: Configurable paths etc., use settings.yml
+    sock.listen();
+    // TODO: Close this using pclose() during client teardown
+    auto result = popen("python2 /home/hudson/puffer/third_party/pensieve/multi_video_sim/rl_test.py /home/hudson/nn_model_ep_77400.ckpt", "r");
```
Comment (francisyyan) [inline src/abr/pensieve.hh]: It's not okay to use raw system calls without checking their return values; at least use `CheckSystemCall`. But in this case, we should use `ProcessManager` (which manages `ChildProcess`).

File: src/abr/pensieve.hh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  double rebuffer_length_coeff_ {REBUFFER_LENGTH_COEFF};
+  double ssim_diff_coeff_ {SSIM_DIFF_COEFF};
+  */
+
+  /* for the current buffer length and quality */
+  size_t curr_buffer_ {};
+  size_t curr_format_ {};
+
+  size_t next_format_ {};
+  static FileDescriptor open_ipc() {
```
Comment (francisyyan) [inline src/abr/pensieve.hh]: Move this function to a member function (without `static`), or simply put it in the constructor.

File: src/abr/pensieve.cc [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    j["next_chunk_sizes"] = next_chunk_sizes_bare; //Bytes
+    uint16_t json_len = j.dump().length();
+    cout << j.dump() << endl;
+
+    connection_.write(put_field2(json_len) + j.dump());
+
+    //Now, busy wait until Pensieve responds (on my laptop this takes between 1 and 4 ms)
+    auto read_data = connection_.read_exactly(2); //read 2 byte length
+    if (read_data.empty()) {
+        cout << "Empty read, ERROR" << endl;
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: This can't happen because `read_exactly` will throw an exception internally.

File: src/abr/pensieve.cc [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+Pensieve::Pensieve(const WebSocketClient & client,
+         const string & abr_name, const YAML::Node & abr_config)
+  : ABRAlgo(client, abr_name)
+{
+
+    cout << abr_config << endl; // TODO: Actually use abr_config or remove it
+    // Config right now is done via constants in the Pensieve file, I am tempted
+    // to leave it that way unless we think we will actually use different
+    // configs for different people/channels.
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: We probably won't change any constants, but you can add and read paths from abr_config to get rid of hardcoded paths.

File: src/abr/pensieve.cc [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        cout << "next_format: " << vformats[next_format_] << endl;
+    }
+}
+
+VideoFormat Pensieve::select_video_format()
+{
+  // Basic design here is that the next_format is set after every video ack,
+  // so any time this is called (which must come after an ack for the previous video chunk) this
+  // value will have been updated.
+  return client_.channel()->vformats()[next_format_];
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: For now, it is fine to assume that all channels always have exactly 10 available bitrates. But notice that the channel might have changed when `select_video_format()` is called, which means that you need to record the `index_in_bitrate_ladder`, and sort the candidates of the next video chunk (of the current channel) by size.

File: src/abr/pensieve.cc [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    size_t vformats_cnt = vformats.size();
+
+    uint64_t next_vts = client_.next_vts().value();
+    const auto & data_map = channel->vdata(next_vts);
+    vector<pair<double, size_t>> next_chunk_sizes; //Store pairs of chunk size, corresponding vf index
+    for (size_t i = 0; i < vformats_cnt; i++) {
+      const auto & vf = vformats[i];
+
+      size_t chunk_size = get<1>(data_map.at(vf));
+      double chunk_size_mb = (double)chunk_size; // Bytes
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: Why not merge this type conversion into the next line? Or simply reply on the implicit conversion?

**Comment by hudson-ayers** [comment]
@francisyyan I think this PR is ready for another review, unless you want me to squash my commits and re-open

File: src/media-server/run_servers.cc [hunk 8]
```diff
@@ -116,6 +116,12 @@ int main(int argc, char * argv[])
     "server_info", "active_streams", "client_buffer", "client_sysinfo",
     "video_sent", "video_acked"};
 
+  /* Remove ipc directory prior to starting Media Server */
+  string ipc_dir = "pensieve_ipc";
+  if (roost::exists(ipc_dir)) {
```
Comment (francisyyan) [inline src/media-server/run_servers.cc]: let's not use `roost` unless `filesystem` can't meet our needs

File: src/net/serialization.cc [hunk 9]
```diff
@@ -0,0 +1,41 @@
+/* -*-mode:c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+#include "serialization.hh"
+
+using namespace std;
+
+string put_field( const uint16_t n )
+{
+  const uint16_t network_order = htobe16(n);
```
Comment (francisyyan) [inline src/net/serialization.cc]: once you decide to put spaces within `( )`, then you need to put them everywhere else (e.g., CS140 will deduct points from your assignments for not being consistent); or you could simply remove all the spaces :-)

File: src/util/pid.cc [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#include "pid.hh"
+#include <exception.hh>
+
+using namespace std;
+
+pid_t pid()
+{
+
```
Comment (francisyyan) [inline src/util/pid.cc]: extra new line

File: src/abr/pensieve.cc [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+using json = nlohmann::json;
+using namespace std;
+
+Pensieve::Pensieve(const WebSocketClient & client,
+                   const string & abr_name, const YAML::Node & abr_config)
+  : ABRAlgo(client, abr_name)
+{
+
+  string ipc_dir = "pensieve_ipc";
+  fs::create_directories(ipc_dir);
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: Better to use https://en.cppreference.com/w/cpp/filesystem/create_directory

File: src/abr/pensieve.cc [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  fs::path ipc_path = fs::current_path() / ipc_file_;
+  string pensieve_path;
+  string nn_path;
+
+  if (abr_config["pensieve_path"] && abr_config["nn_path"]) {
+    pensieve_path = abr_config["pensieve_path"].as<string>();
+    nn_path = abr_config["nn_path"].as<string>();
+  } else {
+    cerr << "Pensieve requires specifying paths in abr_config" << endl;
+    throw "Config Error";
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: need to throw a `runtime_error`

File: src/abr/pensieve.cc [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  pensieve_proc_ = make_unique<ChildProcess>(pensieve_path,
+                                             [&pensieve_path, &prog_args]() {
+      return ezexec(pensieve_path, prog_args);
+    }
+  );
+
+  connection_ =  make_unique<FileDescriptor>(sock.accept());
+}
+
+Pensieve::~Pensieve() {
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: put `{` in the next line

File: src/abr/pensieve.cc [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  error_code ec;
+  if (not fs::remove(ipc_file_, ec)) {
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: don't need `error_code`; simply use `fs::remove(ipc_file_);`

File: src/abr/pensieve.cc [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+  assert(vformats_cnt == 10); // Pensieve requires exactly 10 bitrates
+
+  uint64_t next_vts = client_.next_vts().value();
+  const auto & data_map = channel->vdata(next_vts);
+  vector<double> next_chunk_sizes;
+
+  for (size_t i = 0; i < vformats_cnt; i++) {
+    const auto & vf = vformats[i];
+    double chunk_size = (double)get<1>(data_map.at(vf)); // Bytes
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: nit: `(double)` is redundant

File: src/abr/pensieve.cc [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+  j["playback_buf"] = client_.video_playback_buf(); // Seconds
+  j["rebuf_time"] = client_.cum_rebuffer(); // Seconds
+  j["last_chunk_size"] = (double)size; // Bytes
+  j["next_chunk_sizes"] = next_chunk_sizes; //Bytes
+  uint16_t json_len = j.dump().length();
+
+  connection_->write(put_field(json_len) + j.dump());
+
+  auto read_data = connection_->read_exactly(2); // Read 2 byte length
+  auto rcvd_json_len = get_uint16( read_data.data() );
```
Comment (francisyyan) [inline src/abr/pensieve.cc]: `get_uint16(read_data.data());`

File: src/abr/pensieve.hh [hunk 17]
```diff
@@ -0,0 +1,31 @@
+#ifndef PENSIEVE_HH
+#define PENSIEVE_HH
+
+#include "abr_algo.hh"
+#include "child_process.hh"
+#include "filesystem.hh"
+#include <cstdint>
+#include <cstdlib>
+
+
```
Comment (francisyyan) [inline src/abr/pensieve.hh]: nit: extra new line

**Comment by francisyyan** [comment]
Done my review. Please free to leave "harsh" comments like removing extra new lines to me to resolve (even though they are not actually "harsh"; it's just because we are not using `cpplint`) :-)

**Comment by hudson-ayers** [comment]
Addressed all of your comments in my last commit (and fixed a small bug I found on my own). I also tested it with 7 simultaneous streams for 10 minutes, which worked fine. @francisyyan

**Comment by francisyyan** [comment]
Close this pull request as it was already merged. Thanks!

**Pull request closed without merging**